libs/pika/async_mpi/src/mpi_polling.cpp

```diff
@@ -467,62 +627,175 @@ namespace pika::mpi::experimental {
     {
         if (!s)
         {
             return detail::mpi_data_.default_queues_[0].limit_;
         }
-        PIKA_ASSERT(
-            static_cast<std::uint32_t>(s.value()) <= detail::mpi_data_.default_queues_.size());
-        return detail::mpi_data_.default_queues_[static_cast<std::uint32_t>(s.value())].limit_;
+        PIKA_ASSERT(to_underlying(s.value()) <= detail::mpi_data_.default_queues_.size());
+        return detail::get_stream_ref(s.value()).limit_;
+    }
+
+    // -------------------------------------------------------------
+    void set_max_polling_size(std::size_t p)
+    {
+        detail::mpi_data_.max_polling_requests = p;
+    }
+
+    // -------------------------------------------------------------
+    std::size_t get_max_polling_size()
+    {
+        return detail::mpi_data_.max_polling_requests;
     }
 
-    std::uint32_t get_num_requests_in_flight()
+    // -----------------------------------------------------------------
+    std::size_t get_completion_mode()
     {
-        return detail::mpi_data_.in_flight_;
+        return detail::task_completion_flags_;
     }
 
+    // -----------------------------------------------------------------
+    bool create_pool(
+        pika::resource::partitioner& rp, std::string const& pool_name, pool_create_mode mode)
+    {
+        int is_initialized;
+        MPI_Initialized(&is_initialized);
+        if (is_initialized)
+        {
+            MPI_Comm_rank(MPI_COMM_WORLD, &detail::mpi_data_.rank_);
+            MPI_Comm_size(MPI_COMM_WORLD, &detail::mpi_data_.size_);
+        }
+        else
+        {
+            throw std::runtime_error("MPI must be initialized prior to pool creation");
+        }
+        //
+        int flags = detail::get_completion_mode_default();
+        if (mode == pool_create_mode::force_create)
+            flags |= 1;
+        else if (mode == pool_create_mode::force_no_create)
+            flags &= ~1;
+        else if (mode == pool_create_mode::pika_decides)
+        {
+            // if we have a single rank - disable pool
+            if (detail::mpi_data_.size_ == 1)
+            {
+                flags &= ~1;
+            }
+        }
+        using namespace pika::debug::detail;
+        PIKA_DETAIL_DP(detail::mpi_debug<6>,
+            debug(str<>("completion mode"), bin<8>(flags), detail::mode_string(flags)));
+        // override the variable used to control completion mode and pool flags
+        setenv("PIKA_MPI_COMPLETION_MODE", std::to_string(flags).c_str(), true);
+        // and override the main flag
+        detail::task_completion_flags_ = flags;
+
+        // if pool is now disabled, just exit
+        if ((flags & 1) == 0)
+            return false;
+
+        // Disable idle backoff on the MPI pool
+        using pika::threads::scheduler_mode;
+        auto smode = scheduler_mode::default_mode;
+        smode = scheduler_mode(smode & ~scheduler_mode::enable_idle_backoff);
```

Comment:
This should work (see https://github.com/pika-org/pika/blob/b38f3b6c8fc5ebece3fd730c3b806022645cd583/libs/pika/threading_base/include/pika/threading_base/scheduler_mode.hpp#L72-L76):
```suggestion
        smode = smode & ~scheduler_mode::enable_idle_backoff;
```